# Exercise 1

## Part 1

Loading finnish and multilingual tokenizer to check vocab_size:

In [4]:
from transformers import AutoTokenizer

fintokenizer = AutoTokenizer.from_pretrained("TurkuNLP/bert-base-finnish-cased-v1")
multitokenizer = AutoTokenizer.from_pretrained("google-bert/bert-base-multilingual-cased")

config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

In [ ]:
print(fintokenizer.vocab_size)
print(multitokenizer.vocab_size)

50105
119547


<div class="alert alert-block alert-success">
<b>a)</b> Finnish tokenizer's vocab_size is 50105 tokens, and multilingual tokenizer has just under 200 000 tokens. The multilingual tokenizer though holds tokens from 104 different languages, and is one of the largest. Though, I suspect the token amount per language to be smaller, if every language has the same amount tokens -> (119547/104=1149,49...), only around 1150 tokens are listed to one language. This would mean, that the specified finnish BERT has a better chance to work than the multilangual in finnish language.
</div>

In [16]:
fintokenized: list = []
counter: int = 0

with open("kissa.txt") as f:
    for line in f:
        for word in line.split():
            fintokenized.extend(fintokenizer.tokenize(word))
            counter += 1

print(counter)

print(fintokenized)
print(len(fintokenized))

multitokenized: list = []

with open("kissa.txt") as f:
    for line in f:
        for word in line.split():
            multitokenized.extend(multitokenizer.tokenize(word))

print(multitokenized)
print(len(multitokenized))


145
['Kissa', 'eli', 'kes', '##yk', '##issa', 'tai', 'kotik', '##issa', '(', 'Feli', '##s', 'ca', '##tus', ',', '[', '1', ']', '[', '2', ']', 'aiemmin', 'Feli', '##s', 'sil', '##vest', '##ris', 'ca', '##tus', ')', 'on', 'afr', '##ikan', '##vill', '##ikissa', '##sta', '(', 'Feli', '##s', 'ly', '##bi', '##ca', ')', 'polv', '##e', '##utuva', 'ja', 'peto', '##eläinten', '(', 'Car', '##ni', '##vo', '##ra', ')', 'lahkon', 'kissa', '##eläinten', '(', 'Feli', '##da', '##e', ')', 'heimo', '##on', 'kuuluva', 'kes', '##y', 'nisä', '##käs', '##laji', '.', 'Kissa', '##t', 'ovat', 'suosittuja', 'lemmikki', '##eläimiä', ',', 'ja', 'etenkin', 'maaseudulla', 'ne', 'ovat', 'aina', 'olleet', 'hyödyllisiä', 'hiir', '##ten', 'ja', 'muiden', 'tuho', '##laisten', 'pyyd', '##ystä', '##jinä', '.', 'Ihminen', 'alkoi', 'pitää', 'villi', '##kis', '##soja', 'vilja', '##varasto', '##jen', 'suojeli', '##joina', 'Lähi', '-', 'idässä', 'pian', 'maanviljel', '##yksen', 'keksi', '##misen', 'jälkeen', 'yli', '10', '000',

<div class="alert alert-block alert-success">
<b>b)</b> The Finnish tokenizer tokenizes each word to around 258 tokens when the text file has 145 words. The multilingual tokenizer however tokenized the 145 words to 387 tokens. This was expected, since the multilingual tokenizer tokenizes the words to unnecessary subwords (for example kissa => 'Kiss', '##a') 
</div>

<div class="alert alert-block alert-success">
<b>c)</b> With the openAI GPT-5.x tokenizer visualizer, the same text is tokenized to 392 tokens and with GPT-4 the text is tokenized to 506 tokens. OpenAI uses multilingual models, so a similar type of tokenization results as above was expected, though I expected it to be a tad bit better than the bert tokenizer.
</div>

## Part 2

In [17]:
from transformers import pipeline as pipe

MODEL_NAME = "TurkuNLP/bert-base-finnish-cased-v1"

fin = pipe(task="text-generation", model=MODEL_NAME)

model.safetensors:   0%|          | 0.00/501M [00:00<?, ?B/s]

Cancellation requested; stopping current tasks.


KeyboardInterrupt: 

In [ ]:
print(fin.vocab_size)